File: LeetCode/Medium/807 Max Increase to Keep City Skyline.js
var maxIncreaseKeepingSkyline = function(grid) {
    var west = new Array();
    var north = new Array();
    const n = grid.length;

    for(let i = 0; i < n; i++){
        var wtmp = 0;
        var htmp = 0;
        for(let j = 0; j < n; j++){
            wtmp = Math.max(wtmp, grid[i][j]);
            htmp = Math.max(htmp, grid[j][i]);
        }
        west.push(wtmp);
        north.push(htmp);
    }

    var newGrid = Array.from(Array(n), () => Array(n).fill(0));

    for(let i = 0; i < n; i++){
        for(let j = 0; j < n; j++){
            newGrid[i][j] = Math.min(west[i], north[j]);
        }
    }
    var res = 0;
    for(let x = 0; x < n; x++){
        for(let y = 0; y < n; y++){
            res += (newGrid[x][y] - grid[x][y]);
        }
    }
    return res;
};

File: LeetCode/Medium/807 Max Increase to Keep City Skyline.py
class Solution:
    def maxIncreaseKeepingSkyline(self, grid: List[List[int]]) -> int:
        # 북쪽에서 볼 때
        n = len(grid)
        north = []
        for i in range(n):
            tmp = 0
            for j in range(n):
                tmp = max(tmp, grid[j][i])
            north.append(tmp)

        # 서쪽에서 볼 때
        west = []
        for i in range(n):
            tmp = 0
            for j in range(n):
                tmp = max(tmp, grid[i][j])
            west.append(tmp)
        
        newGrid = [[0] * n for _ in range(n)]
        for i in range(n):
            for j in range(n):
                newGrid[i][j] = min(west[i], north[j])
        
        res = 0
        for i in range(n):
            for j in range(n):
                res += (newGrid[i][j] - grid[i][j])
        
        return res
